readme.md
![](thumbnail.png)

# Bob's & Angel's Proper Pollution

Have you ever wondered why Bob/Angel mid/late game with biters is so easy? Then this mod is made exactly for you! The aim of this mod is to change espacially Bob's energy entities like Boilers and Mining Drills which have **MASSIVELY** reduced pollution production at higher tiers.

# TL;DR

In early game (espacially Death World) the biters keep coming and you need to be on your toes. In later stages of the game, as you upgrade your Boilers/Miners and use modules, you get to the point when you generate almost no pollution and biter attacks basically disappear.
This mod significantly reduces the reduction in pollution for miners/boilers which creates much more consistent experience. You are still rewarded with slight reduction for every new tier you build.

# Examples

Example:
- Mining Drill (Tier 1) has 0.5/s (30/m) mining speed, 10/m polluton and 90kW consumption.
- Mining Drill (Tier 5) has 8.0/s (480/m) mining speed, 2/m polluton and 1.1MW consumption.

So the Tier 1 mining drill generates 0.33 pollution per ore mined, while Tier 5 generates 0.004167 pollution per ore mined, that's **80** times lower.

Of course we need to factor in the energy consumption, which is ~12 times higher. So let's dig into boilers:
- Boiler (Tier 1) produces 1.8 MW of energy at 30/min pollution
- Boiler (Tier 5) produces 9 MW of energy at 9.38/min pollution or 1.8 MW of energy at 1.876/min pollution, which is **16** times less

Even though the Mining Drill (Tier 5) consumes 12 times more energy than Tier 1, it costs 16 times less pollution to produce that energy with Boiler 5.

Let's add the pollution from energy generation to the equation:
- Mining Drill (Tier 1) -  0.5/s (30/m) mining speed, 10/m + 1.5/m from Boiler 1 = 11.5/m pollution => **0.38333333 pollution per ore mined**
- Mining Drill (Tier 5) - 8.0/s (480/m) mining speed, 2/m + 0.23/m from Boiler 5 = 2.23/m pollution => **0.00464583 pollution per ore mined**

All summed up, Mining Drill (Tier 5) produces more than **82.5** times less pollution than Mining Drill (Tier 1) per ore mined.

Since mining & energy generation account for overwhelming majority of pollution AND higher tier biters need more pollution to spawn than small biters, this causes the attacks on your base to taper off in the mid/late game even with crazy Death World settings.

## **Bob's & Angel's Proper Polution** mod to the rescue

Energy generators, reactors and miners now have growing pollution with each tier (it grows slower and slower with higher tiers so the incentive to upgrade is still there).

Let's redo the same example with modded pollution.
- Mining Drill (Tier 1) - unchanged (10/m pollution, 0.5/s mining speed, 90kW consumption)
- Mining Drill (Tier 5) - 83.59/m pollution, rest unchanged (8/s mining speed, 1.1MW consumption)
- Boiler (Tier 1) - unchanged (30/m pollution, 1.8MW)
- Boiler (Tier 5) - 91.521/m pollution, rest unchanged (9MW)

Let's calculate Boilers first:
- Boiler (Tier 1) produces 1.8 MW of energy at 30/min pollution
- Boiler (Tier 5) produces 9 MW of energy at 91.521/min pollution or 1.8 MW of energy at 18.31/min pollution, which is 61% of Boiler Tier 1 or **1.64 times** less

And now the miners:
- Mining Drill (Tier 1) has 0.5/s (30/m) mining speed, 10/m + 1.5/m from Boiler 1 = 11.5/m pollution => **0.38333333 pollution per ore mined** (unchanged)
- Mining Drill (Tier 5) has 8.0/s (480/m) mining speed, 83.59/m + 11.19/m from Boiler 5 = 94.78/m pollution => **0.197457 pollution per ore mined**

That's 51.5% of Tier 1 or **1.94** times less.

Enjoy!

# Playtesting

Tested with Death World setting (300%/300% nests)

Bob's Enemies settings:
- Super spawners - enabled
- Health Increase - enabled
- Bigger enemies soooner - enabled

Bob's Modules settings:
- God modules - disabled
- Green modules - disabled
- Pure productivity modules - disabled
- Pure speed modules - disabled

Bob's Metals, Chemicals and Intermediates settings:
- Expensive electrolysis - enabled

# Special thanks

To [Bobingabout](https://mods.factorio.com/user/Bobingabout) & [Arch666Angel](https://mods.factorio.com/user/Arch666Angel) for creating these amazing mods.



changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 2022.09.18
  Info:
    - Add thumbnail.
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2022.09.18
  Info:
    - Initial release.


data-final-fixes.lua
function SetEmissions(category, entityName, newAmount)
	if not data.raw[category] then return end
	local machine = data.raw[category][entityName]
	
	if not machine then
		log(entityName .. " not found.")
		return
	end
	
	machine.energy_source.emissions_per_minute = newAmount
	log(entityName .. " pollution adjusted.")
end

function SetEmissionsByCoeff(category, entityName, basePollution, coeff)
	if not data.raw[category] then return end
	local machine = data.raw[category][entityName]
	
	if not machine then
		log(entityName .. " not found.")
		return
	end
	
	machine.energy_source.emissions_per_minute = basePollution * coeff
	log(entityName .. " pollution adjusted.")
end

function SetEmissionsByCoeffFluidReactor()

	if not data.raw["reactor"] then return end
	local machine1 = data.raw["reactor"]["fluid-reactor"]
	local machine2 = data.raw["reactor"]["fluid-reactor-2"]
	local machine3 = data.raw["reactor"]["fluid-reactor-3"]
	
	if not machine1 then
		log("fluid-reactor" .. " not found.")
		return
	end
	
	if not machine2 then
		log("fluid-reactor-2" .. " not found.")
		return
	end
	
	if not machine3 then
		log("fluid-reactor-3" .. " not found.")
		return
	end
	
	-- This is needed because util.merge is used when creating fluid-reactor (I guess...?)
	data.raw["reactor"]["fluid-reactor"] = table.deepcopy(machine1)
	data.raw["reactor"]["fluid-reactor-2"] = table.deepcopy(machine2)
	data.raw["reactor"]["fluid-reactor-3"] = table.deepcopy(machine3)
	
	-- Base fluid pollution is 200%
	data.raw["reactor"]["fluid-reactor"].energy_source.emissions_per_minute = 80 * 1.0 / 2
	data.raw["reactor"]["fluid-reactor-2"].energy_source.emissions_per_minute = 120 * 0.95 / 2
	data.raw["reactor"]["fluid-reactor-3"].energy_source.emissions_per_minute = 160 * 0.90 / 2
end

-- Air Filter
SetEmissions("assembling-machine","angels-air-filter",0)
SetEmissions("assembling-machine","angels-air-filter-2",0)

-- Refugiums
SetEmissions("assembling-machine","bio-refugium-puffer",0.6)
SetEmissions("assembling-machine","bio-refugium-biter",0.6)
SetEmissions("assembling-machine","bio-refugium-fish",0.6)

-- Plant Farms
SetEmissions("assembling-machine","crop-farm",-4) -- -40
SetEmissions("assembling-machine","temperate-farm",-8) -- -80
SetEmissions("assembling-machine","swamp-farm",-8) -- -80
SetEmissions("assembling-machine","desert-farm",-8) -- -80

-- Algae Farms
SetEmissions("assembling-machine","algae-farm",-1.67)
SetEmissions("assembling-machine","algae-farm-2",-3.33)
SetEmissions("assembling-machine","algae-farm-3",-5)
SetEmissions("assembling-machine","algae-farm-4",-6.67)
SetEmissions("assembling-machine","bio-generator-temperate-1",-7.5)
SetEmissions("assembling-machine","bio-generator-swamp-1",-7.5)
SetEmissions("assembling-machine","bio-generator-desert-1",-7.5)
SetEmissions("assembling-machine","bio-arboretum-1",-10)

-- Assembling Machines
SetEmissions("assembling-machine","assembling-machine-2",4)
SetEmissions("assembling-machine","assembling-machine-3",5)
SetEmissions("assembling-machine","assembling-machine-4",6)
SetEmissions("assembling-machine","assembling-machine-5",7)
SetEmissions("assembling-machine","assembling-machine-6",8)
SetEmissions("assembling-machine","electronics-machine-1",4)
SetEmissions("assembling-machine","electronics-machine-2",5)
SetEmissions("assembling-machine","electronics-machine-3",6)

-- Centrifuge
SetEmissions("assembling-machine","centrifuge-2",5)
SetEmissions("assembling-machine","centrifuge-3",6)

-- Mining Drills
SetEmissionsByCoeff("mining-drill","electric-mining-drill",10, 1.0)
SetEmissionsByCoeff("mining-drill","bob-mining-drill-1",20, 0.95) -- 1/s
SetEmissionsByCoeff("mining-drill","bob-mining-drill-2",40, 0.90) -- 2/s
SetEmissionsByCoeff("mining-drill","bob-mining-drill-3",80, 0.85) -- 4/s
SetEmissionsByCoeff("mining-drill","bob-mining-drill-4",160, 0.80) -- 8/s

-- Large Area Mining Drills
SetEmissionsByCoeff("mining-drill","bob-area-mining-drill-1",15, 1.0) -- 0.75/s
SetEmissionsByCoeff("mining-drill","bob-area-mining-drill-2",25, 0.95) -- 1.25/s
SetEmissionsByCoeff("mining-drill","bob-area-mining-drill-3",40, 0.90) -- 2/s
SetEmissionsByCoeff("mining-drill","bob-area-mining-drill-4",65, 0.85) -- 3.25/s

-- Pumpjacks
SetEmissionsByCoeff("mining-drill","pumpjack",10, 1.0) -- 1/s
SetEmissionsByCoeff("mining-drill","bob-pumpjack-1",20, 0.95) -- 2/s
SetEmissionsByCoeff("mining-drill","bob-pumpjack-2",30, 0.90) -- 3/s
SetEmissionsByCoeff("mining-drill","bob-pumpjack-3",40, 0.85) -- 4/s
SetEmissionsByCoeff("mining-drill","bob-pumpjack-4",50, 0.80) -- 5/s

-- Boilers
SetEmissionsByCoeff("boiler","boiler",30, 1.0) -- 1.8 MW
SetEmissionsByCoeff("boiler","boiler-2",60, 0.95) -- 3.6 MW
SetEmissionsByCoeff("boiler","boiler-3",90, 0.90) -- 5.4 MW
SetEmissionsByCoeff("boiler","boiler-4",120, 0.85) -- 7.2 MW
SetEmissionsByCoeff("boiler","boiler-5",150, 0.80) -- 9.0 MW

-- Fluid Burning Boilers
-- Base fluid pollution is 200%
SetEmissionsByCoeff("boiler","oil-boiler",80, 0.95 / 2) -- 5.4 MW
SetEmissionsByCoeff("boiler","oil-boiler-2",120, 0.90 / 2) -- 8.1 MW
SetEmissionsByCoeff("boiler","oil-boiler-3",160, 0.85 / 2) -- 10.8 MW
SetEmissionsByCoeff("boiler","oil-boiler-4",200, 0.80 / 2) -- 13.8 MW

-- Fluid Burning Generators
-- Base fluid pollution is 200%
SetEmissionsByCoeff("generator","fluid-generator",20, 0.95 / 2) -- 1.8 MW
SetEmissionsByCoeff("generator","fluid-generator-2",30, 0.90 / 2) -- 2.7 MW
SetEmissionsByCoeff("generator","fluid-generator-3",40, 0.85 / 2) -- 3.6 MW

-- Burner Heat Sources
SetEmissionsByCoeff("reactor","burner-reactor",90, 1.0) -- 5.4 MW
SetEmissionsByCoeff("reactor","burner-reactor-2",120, 0.95) -- 7.2 MW
SetEmissionsByCoeff("reactor","burner-reactor-3",150, 0.90) -- 9.0 MW

-- Fluid Heat Source
SetEmissionsByCoeffFluidReactor()

-- Bob's Electrolyser
SetEmissions("assembling-machine","electrolyser",4)
SetEmissions("assembling-machine","electrolyser-2",5)
SetEmissions("assembling-machine","electrolyser-3",6)
SetEmissions("assembling-machine","electrolyser-4",7)
SetEmissions("assembling-machine","electrolyser-5",8)

-- Bob's Chemical Plant
SetEmissions("assembling-machine","chemical-plant",4)
SetEmissions("assembling-machine","chemical-plant-2",5)
SetEmissions("assembling-machine","chemical-plant-3",6)
SetEmissions("assembling-machine","chemical-plant-4",7)

-- Bob's Oil Refinery
SetEmissions("assembling-machine","oil-refinery",6)
SetEmissions("assembling-machine","oil-refinery-2",8)
SetEmissions("assembling-machine","oil-refinery-3",10)
SetEmissions("assembling-machine","oil-refinery-4",12)

-- Electric Furnaces
SetEmissions("furnace","electric-furnace",1)
SetEmissions("furnace","electric-furnace-2",1.5)
SetEmissions("furnace","electric-furnace-3",2)

-- Electric Furnaces
SetEmissions("assembling-machine","electric-mixing-furnace",1)
SetEmissions("assembling-machine","electric-chemical-mixing-furnace",1.5)
SetEmissions("assembling-machine","electric-chemical-mixing-furnace-2",2)
